clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala

```diff
@@ -42,8 +46,118 @@ class AddExecutor[TI <: TransferIndex](addCommand: AddCommand[TI])
       Future
         .failed(_) logErrorMsg s"Metadata at $location cannot be added to Clio", {
         decoded =>
-          webClient.upsert(addCommand.index)(addCommand.key, decoded)
+          {
+            for {
+              existingMetadata <- if (!addCommand.forceUpdate) {
+                queryForKey(webClient).logErrorMsg(
+                  s"Could not query the $name. No files have been added."
+                )
+              } else {
+                Future.successful(None)
+              }
+              upsertResponse <- addFiles(
+                webClient,
+                decoded,
+                existingMetadata
+              ).fold(
+                left => Future.failed(left).logErrorMsg(left.getMessage),
+                identity
+              )
+            } yield {
+              upsertResponse
+            }
+          }
       }
     )
   }
+
+  private def addFiles(
+    client: ClioWebClient,
+    newMetadata: addCommand.index.MetadataType,
+    existingMetadata: Option[addCommand.index.MetadataType],
+  )(implicit ec: ExecutionContext): Either[Throwable, Future[UpsertId]] = {
+    val differences = existingMetadata.fold(Iterable.empty[(String, Any, Any)]) {
+      diffMetadata(_, newMetadata)
+    }
+    for {
+      upsertResponse <- Either.cond(
+        differences.isEmpty,
+        client
+          .upsert(addCommand.index)(addCommand.key, newMetadata)
+          .logErrorMsg(
+            s"An error occurred while adding the $name record in Clio."
+          ),
+        new RuntimeException(
+          "Adding this document will overwrite the following existing metadata: " +
+            differences
+              .map(
+                diff => s"Field: ${diff._1} Old value: ${diff._3} New value: ${diff._2}"
+              )
+              .mkString("\n") +
+            " Use '--force-update' to overwrite this data."
+        )
+      )
+
+    } yield {
+      upsertResponse
+    }
+  }
+
+  private def diffMetadata(
+    existingMetadata: addCommand.index.MetadataType,
+    newMetadata: addCommand.index.MetadataType
+  ): Iterable[(String, Any, Any)] = {
+    val storedFields =
+      new CaseClassMapper[addCommand.index.MetadataType].vals(existingMetadata)
+    val upsertFields =
+      new CaseClassMapper[addCommand.index.MetadataType].vals(newMetadata)
+
+    val newToOldDiff = upsertFields.toSet.diff(storedFields.toSet).toMap
+    val oldToNewDiff = storedFields.toSet.diff(upsertFields.toSet).toMap
```

Comment:
I'm not sure this'll do what we want... won't it flag fields-only-in-the-server and fields-only-in-new-metadata as errors, instead of just fields-in-both?

Could we do a single set-intersection on the keys, and then diff the values?

Reply:
This is getting the vals and comparing them.. the variables are poorly named.

The only reason we run the new -> old then old -> new diff is so we get values for both old and new to output in the message. But the results are the same.